### PR TITLE
driver/vmwarevsphere: add --vmwarevsphere-folder flag

### LINKED ISSUE #1556: Cannot create VM in specific inventory folder using vmwarevsphere driver
The vSphere environment that I use is very locked down and my user only has permissions for creating VMs in a specific directory. However, it appears that there's no way to control what directory the vmwarevsphere driver uses. Judging from the error it's trying to create a directory with the same name as the docker machine I'm creating.

It would be useful if this worked like the Vagrant vsphere driver where I can specify the "vm_base_path" which is used in combination with the actual name of the VM to construct the path at which the VM is created.

### PR DESCRIPTION
**Pull request by ajfeatherstone**
This commit implements the --vmwarevsphere-folder flag, allowing the
user to specify the folder on the datacenter to create the VM in. The
folder must exist before the VM is created.
Fixes #1556

Signed-off-by: Andrew Featherstone <andrew.featherstone@cambridgeconsultants.com>

### CODE REVIEW

**Comment by SvenDowideit** [comment]
nice!

**Comment by SvenDowideit** [comment]
@frapposelli ?

**Comment by tehkuhnz** [comment]
+1 on this feature

**Comment by yogeek** [comment]
Thank you for this merge. Do you know in which version this feature will be relased please ?

**Comment by ajfeatherstone** [comment]
Is there anything I can do to help get this merged in?

**Comment by yogeek** [comment]
Any news about the merge of this feature please ?

**Comment by MitchellJThomas** [comment]
I'm also interested in this feature.

**Comment by peterlloyd** [comment]
I'm also interested to see this merged

**Comment by cwhits** [comment]
I can't actually get this PR to work. Am I missing something? `govc folder.info /path/to/folder` returns the info on the folder path, but when I use that same path with `--vmwarevsphere-folder` I get 
```
Error with pre-create check: "failed to find VM Folder '/path/to/folder'"
```

**Comment by ajfeatherstone** [comment]
I've not used govc before, @cwhits. Are you able to provide a redacted complete example of the command you've run, e.g. how did you specify the vCenter URL and login credentials?

**Comment by cwhits** [comment]
@ajfeatherstone what parts would you like? I can create a VM just fine if I leave off the `--vmwarevsphere-folder` path. It's only when using that flag do I get the error that it failed to find the VM Folder. Is there a way to echo out what folders it thinks it found?

I tried doing `Folder Name`, `/Datacenter/Folder/Path/Folder Name`, and`/Datacenter/vm/Folder/Path/Folder Name` and none of those worked. They all result in the same `failed to find VM Folder` message.

**Comment by ajfeatherstone** [comment]
I'm interested in how you drive ```govc``` as I've not used that before. I drive Docker Machine through Gitlab's runner, where the folder is specified as ```--vmwarevsphere-folder=Gitlab-CI```. I've pasted a redacted view of vSphere Client v6.0.0. Docker Machine with this PR applied creates the ```runner-``` VMs in this folder.

image

**Comment by cwhits** [comment]
I'm using `govc` and `docker-machine` (with your PR) on macOS 10.12.6.

I don't have privs to list the _entire_ Hosted Customers folder. I can only see a few folders in there. 

vo7ig1sfts

**Comment by ajfeatherstone** [comment]
Thanks for the screenshot, that's very helpful. I think your path in this instance should be ```Hosted Customers/<redacted>/Linux/Testing```, but I haven't tried testing my setup with sub-directories. In my case I also specify the datacenter using the ```--vmwarevsphere-datacenter``` options as well (you can see in the screenshot the vCenter Server has two datcenters in its inventory that I'm permitted to see).

**Comment by cwhits** [comment]
I'm specifying: `vcenter`, `username`, `password`, `datacenter`, `network`, `pool`, `datastore`, and `folder`.

**Comment by ajfeatherstone** [comment]
I think this is the most relevant options I pass to Docker Machine (again, redacted).
```
      --vmwarevsphere-datacenter=redacted-datacenter \
      --vmwarevsphere-datastore=redacted-datastore \
      --vmwarevsphere-folder=Gitlab-CI \
      --vmwarevsphere-pool=/redacted-datacenter/host/redacted-cluster/Resources/gitlab_ci_pool \
      --vmwarevsphere-vcenter=redacted-vcenter-url.com \
```
Can you provide me with a redacted version of your call to ```govc```? I can then try to do the same thing at my end.

**Comment by cwhits** [comment]
for govc I've set env vars for `GOVC_URL`, `GOVC_USERNAME`, `GOVC_PASSWORD`, and `GOVC_DATACENTER`

and then just run `govc folder.info Testing` or `govc folder.info /Redacted Datacenter/vm/Hosted Customers/Redacted/Linux/Testing`

**Comment by ajfeatherstone** [comment]
OK, great thanks. Also sorry this didn't 'just work' for you, as I'd love it to (and for this to get merged upstream). Here's what I get from running ```govc```
```
C:\>govc_windows_amd64.exe folder.info Gitlab-CI
Name:        Gitlab-CI
  Path:      /Redacted Datacenter/vm/Gitlab-CI
  Types:     Folder,VirtualMachine,VirtualApp
  Children:  4
```
So that looks good. Can you confirm that your folder type is the same? If so, then the only obvious difference I can see is that your VMs are created in a sub-directory, and this may be an issue in my implementation as it stands. If so, are you able to try to create VMs in one of your datacenter's 'top level' VM folders? Even if you're not permitted to create the VM, it might be that Docker Machine can at least see the path.
I can try to more closely reproduce your setup (trying to create in a sub-directory), but that probably won't happen until next week.

**Comment by cwhits** [comment]
Yep, my `govc` output matches that
```
Name:        Testing
  Path:      /Redacted Datacenter/vm/Hosted Customers/Redacted/Linux/Testing
  Types:     Folder,VirtualMachine,VirtualApp
  Children:  3
```
If I use the top level folder `Hosted Customers` the pre-create checks all pass (it successfully finds the folder) but I get the error ` Permission to perform this operation was denied.` since I don't have privs to be able to create a VM in that top-level directory.

**Comment by ajfeatherstone** [comment]
Hi @cwhits,

I've replicated the issue at my end and have found the underlying cause. The current implementation of this feature does only work if the VM folder is an immediate child of the datacenter. I'll look at fixing this so that sub-directories are also supported and update this PR.

File: drivers/vmwarevsphere/vsphere.go [hunk 1]
```diff
@@ -321,6 +328,24 @@ func (d *Driver) PreCreateCheck() error {
...
 	f.SetDatacenter(dc)
 
+	// Folder
+	if d.Folder != "" {
+		// Find the specified Folder to create the VM in.
+		folders, err := dc.Folders(ctx)
+		if err != nil {
+			return err
+		}
+		f, err := object.NewSearchIndex(c.Client).FindChild(ctx, folders.VmFolder, d.Folder)
```
Comment (ajfeatherstone) [inline drivers/vmwarevsphere/vsphere.go]: This line is why @cwhits can't use a sub-directory as the destination folder for the VM. As a TODO for me this is what needs fixing (possibly needs to [Folder](https://godoc.org/github.com/vmware/govmomi/find#Finder.Folder)

**Comment by ajfeatherstone** [comment]
Hi @cwhits,

The second commit to this PR allows you to specify a sub-directory to start your VMs in using the form ```Hosted Customers/Redacted/Linux/Testing```. I'm not sure whether this is preferable to ```/Redacted Datacenter/vm/Hosted Customers/Redacted/Linux/Testing```; personally I didn't like the idea of having to specify the redundant information (the datacenter and that the folder is a VM folder). Are you able to take this updated PR and see if it works for you?

**Comment by ajfeatherstone** [comment]
@shin- is there anything I can do to help get this merged?

**Comment by cwhits** [comment]
Yeah, this works for me!

I have other permissions errors preventing me from being able to upload the `boot2docker.iso` but the pre-checks pass and the VM _is actually created_ in the folder that I'm telling it to create it in.

**Comment by cwhits** [comment]
@ajfeatherstone just throwing it out there that `--vmwarevsphere-pool` and `--vmwarevsphere-datastore` both accept the full path `/Datacenter/host/Path/to/Whatever` 
and `/Datacenter/datastore/Path/to/Whatever`, respectively, and the relative path `Path/to/Whatever`.

**Comment by ajfeatherstone** [comment]
I'm glad it's working for you now. I'm fairly ambivalent about the format of paths, but I'm not sure whether supporting multiple path styles is a good idea, i.e. I'd rather go for one form or another. I we were allowing ```/Datacenter/datastore/Path/to/Datastore``` then I'd want to validate that the ```Datacenter``` matches the one specified with the ```--vmwarevsphere-datacenter```.

I'm really keen to hear from the maintainers of this project about what they think about this project, and do what we can to get this PR accepted.

**Comment by ajfeatherstone** [comment]
Hi @shin,

Is there anything we can do to help get this merged onto master? There seems to be plenty of interest in the issue, the CI suggests that this merge won't introduce any regressions, etc. The folk over at https://github.com/docker/docker.github.io/pull/3571 are keen to get this merged in as well, or at least determine whether that PR should be dropped.

**Comment by ajfeatherstone** [comment]
Looks like I've tagged the wrong user, and should have pinged @shin- instead. I'd love to hear your thoughts.

**Comment by dgageot** [comment]
LGTM

**Pull request merged**